File: src/training16/00-hybridMergeSort/mergeSort/mergeSort.py
# Copyright © 2016 Franco Masotti <franco.masotti@student.unife.it>
#                  Danny Lessio
# This work is free. You can redistribute it and/or modify it under the
# terms of the Do What The Fuck You Want To Public License, Version 2,
# as published by Sam Hocevar. See the LICENSE file for more details.


def merge(A, p, q, r):
    n1 = q - p + 1
    n2 = r - q

    print(" p ", p)
    print(" q ", q)
    print(" r ", r)

    L = []
    R = []
    print("A in merge before: ", A)
    for i in range(0, n1):
        L.append( A[ p + i ] )

    for j in range(0, n2):
        R.append( A[ q + 1 + j ])

    L.append( float("inf") )
    R.append( float("inf") )

    print( "L : " , L)
    print( "R : " , R)

    i = 0
    j = 0

    # merging
    for k in range(p, r + 1):
        if L[ i ] <= R[ j ]:
            A[ k ] = L[ i ]
            i = i + 1
        else:
            A[ k ] = R[ j ]
            j = j + 1

    print("A in merge after: ", A)
    print("")

def mergeSort(A, p, r):
    if p < r:
        q = int(( p + r ) / 2) # floor
        mergeSort(A, p, q)
        mergeSort(A, q+1, r)
        merge(A, p, q, r)


if __name__ == "__main__":
    A = [5, 4, 3, 2, 1]
    p = 0
    r = len(A) - 1

    mergeSort(A, p, r)
    print( A )

File: src/training16/00-hybridMergeSort/mergeSort/mergeSort.c
/* Copyright © 2016 Franco Masotti <franco.masotti@student.unife.it>
 * This work is free. You can redistribute it and/or modify it under the
 * terms of the Do What The Fuck You Want To Public License, Version 2,
 * as published by Sam Hocevar. See the LICENSE file for more details.
 */


#ifndef M_MERGESORT_H
#define M_MERGESORT_H
#include "mergeSort.h"
#endif


void mergeSort( int *V, int p, int r)
{

    int q;


    if (p < r)
    {
        q = ( p + r ) / 2;
        mergeSort(V, p, q);
        mergeSort(V, q + 1, r);
        merge(V, p, q, r);
    }


    return;

}

void merge (int *V, int p, int q, int r)
{

    int n1 = q - p + 1;
    int n2 = r - q;
    int i, j, k;
    int *L, *R;


    if ( ( L = malloc( (n1 + 1) * sizeof( int ) ) ) == NULL )
            exit( EXIT_FAILURE );
    if ( ( R = malloc( (n2 + 1) * sizeof( int ) ) ) == NULL )
            exit( EXIT_FAILURE );

    for (i = 0; i < n1; i++)
        L[i] = V[p + i];

    for (j = 0; j < n2; j++)
        R[j] = V[q + 1 + j];

    L[n1] = INF;
    R[n2] = INF;

    i = 0;
    j = 0;
    for (k = p; k <= r; k++)
    {
        if (L[i] <= R[j])
        {
            V[k] = L[i];
            i++;
        }
        else
        {
            V[k] = R[j];
            j++;
        }
    }

    free (L);
    free (R);

    return;

}

